src/System.Linq/tests/ConcatTests.cs

```diff
@@ -89,142 +70,138 @@ public void SecondNull()
+            };
         }
 
-        [Fact]
-        public void ConcatWithSelf()
+        public static IEnumerable<object[]> ConcatWithSelfData()
         {
             IEnumerable<int> source = Enumerable.Repeat(1, 4).Concat(Enumerable.Repeat(1, 5));
             source = source.Concat(source);
-            VerifyEquals(Enumerable.Repeat(1, 18), source);
+
+            yield return new object[] { Enumerable.Repeat(1, 18), source };
         }
 
-        private void VerifyEquals(IEnumerable<int> expected, IEnumerable<int> actual)
+        public static IEnumerable<object[]> ChainedCollectionConcatData() => GenerateSourcesData(innerTransform: e => e.ToList());
+
+        private static IEnumerable<object[]> GenerateSourcesData(EnumerableTransform<int> outerTransform = null, EnumerableTransform<int> innerTransform = null)
         {
-            Assert.Equal(expected, actual);
-            Assert.Equal(expected, actual.ToArray());
-            Assert.Equal(expected, actual.ToList());
-            Assert.Equal(expected, actual.Select(i => i).ToArray());
-            Assert.Equal(expected, actual.Where(i => true).ToArray());
-            Assert.Equal(expected, actual.OrderBy(i => i));
-            Assert.Equal(expected, Enumerable.Empty<int>().Concat(actual));
-            Assert.Equal(expected.Count(), actual.Count());
+            outerTransform = outerTransform ?? (e => e);
+            innerTransform = innerTransform ?? (e => e);
+
+            for (int i = 0; i <= 6; i++)
+            {
+                var expected = Enumerable.Range(0, i * 3);
+                var actual = Enumerable.Empty<int>();
+                for (int j = 0; j < i; j++)
+                {
+                    actual = outerTransform(actual.Concat(innerTransform(Enumerable.Range(j * 3, 3))));
+                }
+
+                yield return new object[] { expected, actual };
+            }
         }
 
-        [Fact]
-        public void ManyEmptyConcats()
+        [Theory]
+        [MemberData(nameof(ManyConcatsData))]
+        public void ManyConcats(IEnumerable<int> toConcat, int times, IEnumerable<int> expected)
         {
+            Debug.Assert(times >= 0);
+
             IEnumerable<int> source = Enumerable.Empty<int>();
-            for (int i = 0; i < 256; i++)
+            for (int i = 0; i < times; i++)
             {
-                source = source.Concat(Enumerable.Empty<int>());
+                source = source.Concat(toConcat);
             }
-            Assert.Equal(0, source.Count());
-            Assert.Equal(Enumerable.Empty<int>(), source);
+
+            Assert.Equal(toConcat.Count() * times, source.Count());
+            VerifyEqualsWorker(expected, source);
+        }
+
+        public static IEnumerable<object[]> ManyConcatsData()
+        {
+            yield return new object[] { Enumerable.Empty<int>(), 256, Enumerable.Empty<int>() };
+            yield return new object[] { Enumerable.Repeat(6, 1), 256, Enumerable.Repeat(6, 256) };
```

Comment:
Also, by using the same value over and over, your change loses some of the verification in the original, e.g. if the concat messed up the ordering, the previous version of the test would catch it, but your change wouldn't.